### PR TITLE
Add configurable warning for lv version mismatch between client and server

### PR DESCRIPTION
**Pull request by Ian-GL**
This PR warns on a version mismatch between the client and server versions of LiveView, as reported on issue #929 . The client sends its version to the server as part on the connection and as part of the mounting process it is checked against the server version. This results in a repeated warning every time there is a new socket connection, I decided to go this way since it was really easy to pass the version as an extra connection param.

This is can also be turned off by adding `config :phoenix_live_view, :warn_version_mismatch, false` on the configuration of the project using LV. If no config found, it defaults to true.

In the PR I only changed the readable `live_view.js`, not sure how the minified version should be updated in a PR since master could change at any time. Open to comments if there is a preferred way to do the warning.

### CODE REVIEW

File: assets/js/phoenix_live_view.js
Comment (josevalim) [inline assets/js/phoenix_live_view.js:12]: Does this bring the whole .json file as a dependency? I am also wondering if we should use the package version or use a separate version scheme for the wire protocol. :)

File: lib/phoenix_live_view/channel.ex
Comment (josevalim) [inline lib/phoenix_live_view/channel.ex:602]: It is preferable to this:

```suggestion
    server_version = Application.spec(:phoenix_live_view :vsn) |> List.to_string()
```

As Mix is not available during a release. :) Although we should decide if we want to use the package version or a separate one.

File: assets/js/phoenix_live_view.js
Comment (Ian-GL) [reply assets/js/phoenix_live_view.js:12]: Yes, it does bring the whole json file. I think it is fine to use a separate version scheme as long as it is also updated on every release. Don't know how you feel about having to update versioning on an extra file, but would be good so as not to import the package json file. :)

File: lib/phoenix_live_view/channel.ex
Comment (Ian-GL) [reply lib/phoenix_live_view/channel.ex:602]: I didn't know that, thanks for pointing it out. When talking about this separate versioning, would it be in a file that is consumed both by the server code and the client one, or also set in two different places?

**Comment by Ian-GL** [comment]
Seeing how the tests failed, the version expected should also be included in tests so it can match with all the params sent from the connection.

File: lib/phoenix_live_view/channel.ex
Comment (henrik) [inline lib/phoenix_live_view/channel.ex:609]: This assumes it's the client side that needs updating. Could it ever be the case that you've updated the client-side and the server side falls behind?

Comment (josevalim) [reply lib/phoenix_live_view/channel.ex:609]: Regardless if the client is behind or ahead, "syncing" the version uses the same command. :)

Comment (henrik) [reply lib/phoenix_live_view/channel.ex:609]: Oh yeah, I forgot it's a file dependency pointing to `../deps` rather than some published package :thumbsup:

File: assets/js/phoenix_live_view.js
Comment (ccblaisdell) [reply assets/js/phoenix_live_view.js:12]: This issue is a bit old, but it looks like there are ways webpack can [bake in a version number at compile time](https://github.com/webpack/webpack/issues/237) without importing `package.json`

**Comment by Ian-GL** [comment]
Sorry for taking so long to follow-up on this. I made the change that @ccblaisdell suggested by injecting the const via webpack, so now webpack is consuming the `package.json` file, but it is not part of the final build. As a side effect of this I had to add a constant with the same name to the jest configuration so the tests don't crash with an undefined const.

In my opinion, given that this does not expose the json file I think it would be good to just leave the comparison between the already defined versions, especially so it doesn't add more places to keep track of every time the version is bumped.

Open to your suggestions.

File: assets/package.json
Comment (josevalim) [inline assets/package.json:35]: Shouldn't this be the same version as the package version? Or am I misunderstanding what "test_mock_version" is? :)

File: lib/phoenix_live_view/channel.ex
Comment (josevalim) [inline lib/phoenix_live_view/channel.ex:621]: You can provide a default value for the environment by adding `env: [warn_version_mismatch: true]` here:

https://github.com/phoenixframework/phoenix_live_view/blob/master/mix.exs#L35

And then here you can do:

```suggestion
    should_display_warning = Application.fetch_env!(:phoenix_live_view, :warn_version_mismatch)
```

File: lib/phoenix_live_view/channel.ex
Comment (josevalim) [inline lib/phoenix_live_view/channel.ex:618]: You can collect the version at compile time with this?

    @version Mix.Project.config()[:version]

My suggestion though is to match only on the prefix, so I would do something like this:

    @version Mix.Project.config()[:version]
    @prefix_version String.split(@version, ".") |> Enum.drop(-1) |> Enum.join(".") |> Kernel.<>(".")

And then inside you do:

    if should_display_warning? and is_string(client_version) and not String.starts_with?(client_version, @prefix_version) do

That's because 0.14.2 works with 0.14.1 but there is no guarantee it works with 0.15 or 0.13. WDYT?

File: lib/phoenix_live_view/channel.ex
Comment (josevalim) [inline lib/phoenix_live_view/channel.ex:626]: I believe we should include the name of the LiveView we are trying to execute here. That's because sometimes the client is embedded externally, such as in the LiveView Dashboard. So it would be good to know if the error is coming from an external application or our own application. WDYT?

File: assets/package.json
Comment (Ian-GL) [reply assets/package.json:35]: Hi Jose, sorry for taking so long to respond. So, following the suggestion made by @ccblaisdell , the version on the JS side is injected via webpack when building and is exposed in the code as a global var. In this case the tests that directly call the js code don't have this treatment, so this is a way of exposing the global constant. This could be the manually set to the same one as the real version, but it could easily get out of date, since instead of testing if it is the same one it would only test that the var defined here is the same as in the result of the executed js function.

File: lib/phoenix_live_view/channel.ex
Comment (Ian-GL) [reply lib/phoenix_live_view/channel.ex:618]: Hmm, on one hand I think that makes sense since it is always guaranteed to work and not break, on the other hand, i think this feature is also really useful to track why something might not be working, ie: with the `push_event` changes, maybe the dev using it updates the mix part but forgets the npm part, so calling `push_event` won't break, but the js client won't get the call because it has not been updated yet. Keeping it specific I think would help with this type of problems when one might spend some time debugging only to find out it was a mismatch error.

File: lib/phoenix_live_view/channel.ex
Comment (Ian-GL) [reply lib/phoenix_live_view/channel.ex:626]: Yes, on this one I completely agree, the more specific this is the better, imo. I'm not 100% sure how to do this but will give it a try.

**Comment by josevalim** [comment]
Hi @Ian-GL, can you please rebase your branch on master? At the moment it has mixed commits from other branches which makes it a bit hard to review.

Thank you so much! :heart:

**Comment by Ian-GL** [comment]
Hi @josevalim . Just rebased and pushed again! Thank you for reviewing this PR! :)

**Review by josevalim — APPROVED** [review]
LGTM!

**Comment by josevalim** [comment]
Actually, we may need to validate that it is within the same minor version. :) Otherwise the warnings would be too frequent. But we can do it in a later pass after merge.

File: assets/js/phoenix_live_view.js
Comment (chrismccord) [inline assets/js/phoenix_live_view.js:1582]:
```suggestion
        vsn: this.parent ? undefined : PACKAGE_VSN
```

We should only send the version on the root LiveViews to avoid sending unnecessary data and checks. `undefined` values are automatically stripped form the JSON payload in JS land. We will also need to conditional check for the vsn param in the elixir side. Note I renamed `lv_version` to `vsn` as well. Thanks!

File: assets/test/view_test.js
Comment (josevalim) [inline assets/test/view_test.js]:
```suggestion
      "flash": undefined, "params": { "_mounts": 0 }, "session": "abc123", "static": null, "url": undefined, "vsn": PACKAGE_VSN}
```

File: assets/test/view_test.js
Comment (josevalim) [inline assets/test/view_test.js]:
```suggestion
      "flash": undefined, "session": "abc123", "static": null, "url": undefined, "vsn": PACKAGE_VSN,
```

File: lib/phoenix_live_view/channel.ex
Comment (josevalim) [inline lib/phoenix_live_view/channel.ex:580]: phx_socket is not the same as lv_socket, so I think this will either always return true or return false. I think the best we can do for now is to check the version match only if the "vsn" parameter has been given.

File: assets/webpack.config.js
Comment (josevalim) [inline assets/webpack.config.js:34]: Actually, I had an idea. Can we make it so PACKAGE_VSN here is only the first two parts? I.e. instead of "0.14.1", it will be "0.14". I was thinking about doing it on the server but I think we can precompute it both on the client and server.

**Comment by chrismccord** [comment]
Thank you for this work and holding tight! This is no longer necessary as we have dropped webpack and the esbuild approach will avoid having to have the end-user blow away node modules on LV updates. Thanks!

**Pull request closed without merging**
